### PR TITLE
C++: Add tests for experimental `cpp/guarded-free` query

### PR DESCRIPTION
**Pull request by jketema**
I investigated a bit if the query can be promoted (instead of moving it to a separate repo). The tests highlight some issues.

#### All query authors

- [ ] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [ ] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+#define TRY_FREE(x) \
+  if (x) free(x);
+
+void test11(char *x) {
+  TRY_FREE(x) // BAD
+}
+
+bool test12(char *x) {
+  if(!x) // GOOD [FALSE POSITIVE]: return value depends on x
```
Comment (paldepind) [inline cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:98]:
```suggestion
  if (!x) // GOOD [FALSE POSITIVE]: return value depends on x
```

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    return false;
+
+  free(x);
+  return true;
+}
+
+void test13(char *x) {
+  if(x != nullptr) // BAD
```
Comment (paldepind) [inline cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:106]:
```suggestion
  if (x != nullptr) // BAD
```

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    free(x);
+}
+
+void inspect(char *x);
+
+void test14(char *x) {
+  if(x != nullptr) // GOOD [FALSE POSITIVE]: x might be accessed
```
Comment (paldepind) [inline cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:113]:
```suggestion
  if (x != nullptr) // GOOD [FALSE POSITIVE]: x might be accessed
```

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 1 above)
Comment (paldepind) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:98]: Just to be consistent with the `if` statements above.

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    inspect(x), free(x);
```
Comment (paldepind) [inline cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:114]: Is the comma operator what's causing the issue? Or does this also fail with `{ inspect(x); free(x); }? If it does then maybe just use semicolons? If not, would it make sense to add a comment to clarify that?

**Review by paldepind — CHANGES_REQUESTED** [review]
Looks good. Just a few comments 😄

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 4 above)
Comment (jketema) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:114]: The problem is that `inspect` may assume that `x` is non-null, so if the remove the `if`, which is what the query is about, the thing would start crashing. Note that this derives from code I saw in the wild. Replacing it by `{ inspect(x); free(x); }` would not be an FP.

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 1 above)
Comment (jketema) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp]: Fixed.

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 2 above)
Comment (jketema) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp]: Fixed.

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 3 above)
Comment (jketema) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp]: Fixed.

File: cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp (hunk 4 above)
Comment (jketema) [reply cpp/ql/test/experimental/query-tests/Best Practices/GuardedFree/test.cpp:114]: Done.

**Review by paldepind — APPROVED** [review]

**Pull request merged**